### ISSUE TITLE
transform produces different result if called with AST

### ISSUE THREAD

**Issue by overlookmotel** [swc-project/swc#5168]
### Describe the bug

I'm not sure if this is a bug or not, but it's certainly strange.

```js
const { transformSync, parseSync } = require('@swc/core');

const code = 'x ??= true';
console.log( transformSync( code ) );
console.log( transformSync( parseSync( code ) ) );
```

Output:

```js
x !== null && x !== void 0 ? x : x = true;
x = x !== null && x !== void 0 ? x : true;
```

The first is more correct. If `x` is a global defined with a setter, the 2nd will trigger the setter even when `x` is not nullish - which `x ??= true` doesn't.

```js
Object.defineProperty( globalThis, 'x', {
  get(v) { return true; },
  set(v) { console.log('set!'); }
} );

x ??= true; // Nothing happens
x = x !== null && x !== void 0 ? x : true; // Logs "set!"
```

But mainly I raise this because it may indicate some other problem. Perhaps some transforms are being skipped, or the wrong target is being selected.

Any idea what might be causing this?

### Input code

```typescript
x ??= true
```

### Config

```json
n/a
```

### Expected behavior

Output in both cases should be `x !== null && x !== void 0 ? x : x = true;`.

### Version

1.2.211

**Comment by overlookmotel** [comment]
Here's another one:

```js
const code = 'c ??= class {}';
console.log( transformSync( code ) );
console.log( transformSync( parseSync( code ) ) );
```

Output:

```js
function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
c !== null && c !== void 0 ? c : c = function c() {
    "use strict";
    _classCallCheck(this, c);
};

function _classCallCheck(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
c = c !== null && c !== void 0 ? c : function _class() {
    "use strict";
    _classCallCheck(this, _class);
};
```

The 2nd case is a regression of #5102.

**Comment by magic-akari** [comment]
There is something wrong with the `deserialize_json`.

**Comment by magic-akari** [comment]
There are two ways to get Rust side AST.

source code -> parse -> AST
AST JSON(parsed from source code) -> deserialize -> AST

Take the following code as an example.

```JavaScript 
x ??= input;
```

It will be parsed into the following AST.

```
AssignExpr {
    span: Span {..},
    op: "??=", 
    left: Expr(Ident(Ident { span: Span {..}, sym: Atom('x' type=inline), optional: false })), 
    right: Lit(Bool(Bool { span: Span {..}, value: true }
}
```
And AST JSON:

```json
{
  "type": "AssignmentExpression",
  "span": "{..}",
  "operator": "??=",
  "left": {
    "type": "Identifier",
    "span": "{..}",
    "value": "x",
    "optional": false
  },
  "right": {
    "type": "BooleanLiteral",
    "span": "{..}",
    "value": true
  }
}
```

But the AST JSON will be deserialize into the following AST.

```
AssignExpr {
    span: Span {..},
    op: "??=", 
    left: Pat(Ident(BindingIdent { id: Ident { span: Span {..}, sym: Atom('x' type=inline), optional: false }, type_ann: None })), 
    right: Lit(Bool(Bool { span: Span {..}, value: true 
}
```

Can we adjust the deserialization logic like the parser?

https://github.com/swc-project/swc/blob/b87e3b0d4f46e6aea1ee7745f0bb3d129ef12b9c/crates/swc_ecma_parser/src/parser/pat.rs#L602-L610

**Comment by overlookmotel** [comment]
@magic-akari Thank you, you're right!

I've submitted PR #5179 to fix. However, my knowledge of Rust is very basic. Could you possibly have a look and see if there's a better way to write this?

**Comment by overlookmotel** [comment]
Fixed in #5179.
